### ISSUE TITLE
Kusama Node Fails to Sync Beyond 8949248

### ISSUE THREAD

**Issue by joepetrowski** [paritytech/polkadot#3778]
A Kusama archive node has failed to sync beyond block 8949248. It often sees that there is a higher block and says "Syncing 0.0 bps", but then reverts back to "Idle".

Have resynced from genesis and tried restarting the node.

Options used: `./polkadot --no-prometheus --no-telemetry --database=paritydb-experimental --pruning=archive --wasm-execution=Compiled --chain=kusama -lsync=trace`

Log (from genesis): https://drive.google.com/file/d/1zY6T2YCn8K6yQ0fWLkY3lBYfRJnnkg7o/view?usp=sharing

Log (on restart): https://drive.google.com/file/d/1UUtR-p6Y0NgkUzaJSZ7mvTslzgjdrSr_/view?usp=sharing

cc @arkpar

**Comment by arkpar** [comment]
Looks like as as soon as the node sends a block request to a peer it disconnects. I.e
```
2021-09-03 10:33:02.437 TRACE tokio-runtime-worker sync: New block request for 12D3KooWLTqCsfhf1Ur71LBCP91yE5YoBdQHsfU5nb5kp2c5Hs1J, (best:9061641, common:8949476) BlockRequest { id: 0, 
...
2021-09-03 10:33:02.712 DEBUG tokio-runtime-worker sync: 12D3KooWLTqCsfhf1Ur71LBCP91yE5YoBdQHsfU5nb5kp2c5Hs1J disconnected     
```
Could you also add `-l sub-libp2p=trace` and post logs again?

cc @tomaka

**Comment by joepetrowski** [comment]
Libp2p logs attached.
[kusama-parity-db-sync-restart-libp2p.log](https://github.com/paritytech/polkadot/files/7105756/kusama-parity-db-sync-restart-libp2p.log)

**Comment by DanielMagro97** [comment]
We're having the same issue on a node, it's also a Kusama archive node but we're using RocksDB instead of ParityDB.
We also have a list of reserved-nodes so it's weird that the connection looks like it  drops.

**Comment by xlc** [comment]
Parachains are having the same issue. So this is a just a polkadot issue not related to parachain setup. https://github.com/paritytech/polkadot/issues/3756

**Comment by tomaka** [comment]
Could it be that the block is too large for the hardcoded limit of the protocol?

**Comment by tomaka** [comment]
Yes:

> 2021-09-03 14:08:45.146 DEBUG tokio-runtime-worker sub-libp2p: Libp2p => Disconnected(PeerId("12D3KooWFZxUeWA4G7aY1FzZLFazcLHphwB1XR92jEj1mLYsGp8b"), Some(Handler(Handler(B(Upgrade(Apply(Custom { kind: InvalidInput, error: "Response size exceeds limit: 23814046 > 16777216" })))))))

**Comment by joepetrowski** [comment]
Is that a block? My validator nodes are running fine, it's just a full node that has the issue. Why do some nodes advance past this but not others?

**Comment by ioannist** [comment]
I am trying to sync a Moonriver collator node from scratch (has to be archive mode) and I have been trying to work around this for a week. Things I have tried and failed:

- Tried 2 different machines, 2 times each
- Tried to sync the relay chain using older versions of Kusama and upgrading as needed
- Tried moonbeam 0.9.6

**Comment by ioannist** [comment]
> Is that a block? My validator nodes are running fine, it's just a full node that has the issue. Why do some nodes advance past this but not others?

Happens when archive mode is on, and you are trying to sync from scratch (or perhaps, starting before #8949248)

**Comment by rphmeier** [comment]
https://kusama.subscan.io/block/8949248?tab=event I don't see anything in the block that would lead it to have such a large size.

**Comment by arkpar** [comment]
Not a single block, but rather a response containing multiple blocks is oversized. The code that should have limited the number of blocks in the response contains a bug.

**Comment by atenjin** [comment]
@bkchr 
It seems this issue have not been solved

follow from issue https://github.com/paritytech/polkadot/issues/3791, and I checkout the commit https://github.com/paritytech/polkadot/commit/1661043836723267562c31e54e844bbc42f6d9fa
which has contains the substrate commit https://github.com/paritytech/substrate/pull/9692

But it still stuck on 
```bash
Sep 09 12:22:43 node1 kusama[34900]: 2021-09-09 12:22:43 💤 Idle (24 peers), best: #8949641 (0x135a…dd8f), finalized #8949248 (0x102e…0032), ⬇ 56.8kiB/s ⬆ 8.9kiB/s
```

```bash
-># ./kusama --version
polkadot 0.9.10-9f70f94cc-x86_64-linux-gnu
```

**Comment by sashman-jaspin** [comment]
So, any workaround to this?  We had two archive nodes syncing and both got stuck on different blocks mentioned here.  According to telemetry there are 100s of other nodes stuck on 8,949,573 and 8,949,641, which we are also stuck on.  

polkadot 0.9.9-1-e3901b89c-x86_64-linux-gnu

**Comment by GopherJ** [comment]
all of our collators got stuck at 8949248

**Comment by tomaka** [comment]
If you could please provide logs with `-l sync=debug -l sub-libp2p=debug`

**Comment by arkpar** [comment]
64 block request is still too much. The fix needs to be deployed on a significant number of peers to work properly. 
@paritytech/ci could we deploy current `0.9.10` on our bootnodes please?

**Comment by GopherJ** [comment]
I have some logs like:

![image](https://user-images.githubusercontent.com/33961674/133199851-dcb8ef0c-4214-489a-a905-3de2feb085ba.png)

**Comment by bkchr** [comment]
> 64 block request is still too much. The fix needs to be deployed on a significant number of peers to work properly.
> @paritytech/ci could we deploy current `0.9.10` on our bootnodes please?

@arkpar but isn't the requesting node decides on how many blocks it wants to request?

**Comment by arkpar** [comment]
@bkchr Requesting side specifies the maximum count, but the serving side applies the size limit. I've changed the requested size from 128 to 64 blocks but the result is still above the limit so that serving nodes that are not updated with the fix yet can't send it. I don't think we should further reduce the requested block size as it will result in too fragmented batches for small-sized blocks. Instead we should deploy the fix on the bootnodes to respect the size limit.

**Comment by adastreamer** [comment]
Hello,

How this issue can be resolved?

**Comment by adastreamer** [comment]
Why the issue is marked as resolved if it's actually not?

**Comment by ordian** [comment]
It's resolved on master and will be fixed in the next release of polkadot.

**Comment by adastreamer** [comment]
> It's resolved on master and will be fixed in the next release of polkadot.

Just built from master, and getting this:
![2021-09-18 15 47 26](https://user-images.githubusercontent.com/919292/133889276-80e4de41-bd3f-42de-8b46-52179dabe2b5.jpg)

Nothing is synced at all :-(

**Comment by adastreamer** [comment]
Removing chain and full resync doesn't help also

**Comment by h4x3rotab** [comment]
Khala Network is meeting the exact same problem. Nice to see the solution was merged, but we are running a mining launch process where thousands of nodes are syncing the Kusama blockchain.

Is the fix included in v0.9.10?

**Comment by bkchr** [comment]
> Khala Network is meeting the exact same problem. Nice to see the solution was merged, but we are running a mining launch process where thousands of nodes are syncing the Kusama blockchain.
> 
> Is the fix included in v0.9.10?

Yes it is included.
